jetty-servlet/src/test/java/org/eclipse/jetty/servlet/DefaultServletTest.java

```diff
@@ -813,6 +815,46 @@ public void testWelcomeMultipleBasesBase() throws Exception
         }
     }
 
+    @Test
+    public void testIncludedWelcomeDifferentBase() throws Exception
+    {
+        Path altRoot = workDir.getPath().resolve("altroot");
+        FS.ensureDirExists(altRoot);
+        Path altIndex = altRoot.resolve("index.html");
+
+        ServletHolder defholder = context.addServlet(DefaultServlet.class, "/alt/*");
+        defholder.setInitParameter("resourceBase", altRoot.toUri().toASCIIString());
+        defholder.setInitParameter("dirAllowed", "false");
+        defholder.setInitParameter("redirectWelcome", "false");
+        defholder.setInitParameter("welcomeServlets", "true");
+        defholder.setInitParameter("pathInfoOnly", "true");
+
+        ServletHolder gwholder = new ServletHolder("gateway", new HttpServlet() {
+            @Override
+            protected void doGet(HttpServletRequest req, HttpServletResponse resp)
+                    throws ServletException, IOException {
+                req.getRequestDispatcher("/alt/").include(req, resp);
+            }
+        });
+        context.addServlet(gwholder, "/gateway/*");
+
+        String rawResponse;
+        HttpTester.Response response;
+
+        // Test included alt default
+        rawResponse = connector.getResponse("GET /context/gateway HTTP/1.0\r\n\r\n");
```

Comment:
This should be `GET /context/gateway/ HTTP/1.0\r\n\r\n` as you are requesting content belonging to the url-pattern of `/gateway/*` which the request to `/gateway` will not match.

Reply:
I checked `org.eclipse.jetty.http.pathmap.PathMappings#getMatch()` and `/p` request is successfully matched for servlet mapped to `/p/*` - it's kept in `org.eclipse.jetty.http.pathmap.PathMappings#_prefixMap`:
```
_prefixMap = {org.eclipse.jetty.util.ArrayTernaryTrie@2636} "{/p=MappedResource[pathSpec=ServletPathSpec["/p/*",pathDepth=2,group=PREFIX_GLOB],resource=default-servlet@2c16bab9==org.ops4j.pax.web.service.jetty.internal.EmbeddedJettyTest$11,jsp=null,order=-1,inst=true,async=true]}"
 LO: int  = 1 (0x1)
 EQ: int  = 2 (0x2)
 HI: int  = 3 (0x3)
 ROW_SIZE: int  = 4 (0x4)
 _tree: char[]  = {char[512]@2893} 
 _key: java.lang.String[]  = {java.lang.String[128]@2894} 
  2 = "/p"
 _value: java.lang.Object[]  = {java.lang.Object[128]@2895} 
  2 = {org.eclipse.jetty.http.pathmap.MappedResource@2643} "MappedResource[pathSpec=ServletPathSpec["/p/*",pathDepth=2,group=PREFIX_GLOB],resource=default-servlet@2c16bab9==org.ops4j.pax.web.service.jetty.internal.EmbeddedJettyTest$11,jsp=null,order=-1,inst=true,async=true]"
   pathSpec: org.eclipse.jetty.http.pathmap.PathSpec  = {org.eclipse.jetty.http.pathmap.ServletPathSpec@2897} "ServletPathSpec["/p/*",pathDepth=2,group=PREFIX_GLOB]"
   resource: java.lang.Object  = {org.eclipse.jetty.servlet.ServletHolder@2656} "default-servlet@2c16bab9==org.ops4j.pax.web.service.jetty.internal.EmbeddedJettyTest$11,jsp=null,order=-1,inst=true,async=true"
```

Reply:
I think you've got 2 differents tests to do here:

1. url is /context/gateway  WITHOUT the trailing slash, which will go to line 377 of ResourceService that does a 302 redirect to url that adds the trailing slash

2. url is /context/gateway/  WITH the trailing slash, which will go to the code at line 402 of ResourceService

Reply:
@grgrzybek I'd still like to see one request with `GET /context/gateway` and one with `GET /context/gateway/`, because the first one will do a redirect before doing the new include path handling: I just want to ensure that we've exercised both codepaths.